crates/server/src/store/blocklist/authority.rs

```diff
@@ -0,0 +1,689 @@
+/// to do so - it is both more efficient, and more secure, to allow the blocklist to drop queries
+/// pre-emptively, as in the first example.
+pub struct BlocklistAuthority {
+    origin: LowerName,
+    blocklist: HashMap<LowerName, bool>,
+    wildcard_match: bool,
+    min_wildcard_depth: u8,
+    sinkhole_ipv4: Ipv4Addr,
+    sinkhole_ipv6: Ipv6Addr,
+    ttl: u32,
+    block_message: Option<String>,
+    consult_action: BlocklistConsultAction,
+}
+
+impl BlocklistAuthority {
+    /// Read the Authority for the origin from the specified configuration
+    pub async fn try_from_config(
+        origin: Name,
+        _zone_type: ZoneType,
+        config: &BlocklistConfig,
+        base_dir: Option<&Path>,
+    ) -> Result<Self, String> {
+        info!("loading blocklist config: {origin}");
+
+        let sinkhole_ipv4 = Ipv4Addr::from_str(&config.sinkhole_ipv4[..]).unwrap_or_else(|_| {
+            info!(
+                "invalid IPv4 sinkhole address '{}' provided; using default 0.0.0.0",
+                &config.sinkhole_ipv4[..]
+            );
+            Ipv4Addr::new(0, 0, 0, 0)
+        });
+        let sinkhole_ipv6 = Ipv6Addr::from_str(&config.sinkhole_ipv6[..]).unwrap_or_else(|_| {
+            info!(
+                "invalid IPv6 sinkhole address '{}' provided; using default ::0",
+                &config.sinkhole_ipv6[..]
+            );
+            Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 0)
+        });
+
+        let mut authority = Self {
+            origin: origin.into(),
+            blocklist: HashMap::new(),
+            wildcard_match: config.wildcard_match,
+            min_wildcard_depth: config.min_wildcard_depth,
+            sinkhole_ipv4,
+            sinkhole_ipv6,
+            ttl: config.ttl,
+            block_message: config.block_message.clone(),
+            consult_action: config.consult_action,
+        };
+
+        let base_dir = base_dir
+            .unwrap_or_else(|| {
+                error!("invalid blocklist (zone directory) base path specified: {base_dir:?}");
+
+                // If no path was specified, try looking relative to the server's current directory
+                // for the blocklists.
+                Path::new("./")
+            })
+            .display();
+
+        // Load block lists into the block table cache for this authority.
+        for bl in &config.lists {
+            info!("adding blocklist {bl}");
+
+            match File::open(format!("{base_dir}/{bl}")) {
+                Ok(handle) => {
+                    if let Err(e) = authority.add(handle) {
+                        error!(
+                            "unable to add data from blocklist {base_dir}/{bl} to blocklist: {e:?}"
+                        );
+                    }
+                }
+                Err(e) => error!("unable to open blocklist file {base_dir}/{bl}: {e:?}"),
+            }
+        }
+
+        Ok(authority)
+    }
```

Comment:
Same for this stuff. If the configuration is wrong, we should fail to initialize.